src/esp/geo/VoxelGrid.cpp

```diff
@@ -0,0 +1,923 @@
+
+#include <assert.h>
+#include <limits.h>
+#include <cmath>
+
+#include <Corrade/Containers/ArrayViewStl.h>
+#include <Magnum/GL/DefaultFramebuffer.h>
+#include <Magnum/MeshTools/Interleave.h>
+#include <Magnum/MeshTools/Reference.h>
+#include <Magnum/Primitives/Cube.h>
+
+#include "VoxelGrid.h"
+#include "esp/assets/ResourceManager.h"
+
+namespace Mn = Magnum;
+namespace Cr = Corrade;
+
+namespace esp {
+namespace geo {
+
+#ifdef ESP_BUILD_WITH_VHACD
+VoxelGrid::VoxelGrid(const std::unique_ptr<assets::MeshData>& meshData,
+                     int resolution) {
+  VHACD::IVHACD* interfaceVHACD = VHACD::CreateVHACD();
+  /*VHACD::IVHACD::Parameters params;
+  params.m_resolution = resolution;
+  params.m_oclAcceleration = false;*/
+  Mn::Debug() << "Voxelizing mesh..";
+
+  // run VHACD
+  interfaceVHACD->computeVoxelField(&meshData->vbo[0][0], meshData->vbo.size(),
+                                    &meshData->ibo[0], meshData->ibo.size() / 3,
+                                    resolution);
+
+  // get VHACD volume, set scale and dimensions
+  VHACD::Volume* vhacdVolume = interfaceVHACD->getVoxelField();
+  double scale = vhacdVolume->getScale();
+  m_voxelSize = Mn::Vector3(scale, scale, scale);
+  const size_t* dims = vhacdVolume->getDimensions();
+  m_voxelGridDimensions = Mn::Vector3i(dims[0], dims[1], dims[2]);
+
+  VHACD::Vec3<double> center = vhacdVolume->getCenter();
+
+  // VHACD computes a axis-aligned bounding box; we need to offset the voxelgrid
+  // by the minimum corner of the AABB
+  VHACD::Vec3<double> minBB = vhacdVolume->getMinBB();
+  m_offset = Mn::Vector3(minBB[0], minBB[1], minBB[2]);
+
+  // create empty VoxelGrid
+  const int gridSize = dims[0] * dims[1] * dims[2];
+  bool* b_grid = new bool[gridSize];
+  std::shared_ptr<bool> boundary_grid(b_grid);
+  boolGrids_.insert(std::make_pair("Boundary", boundary_grid));
+  int num_filled = 0;
+  // Transfer data from Volume to VoxelGrid
+  for (int i = 0; i < dims[0]; i++) {
+    for (int j = 0; j < dims[1]; j++) {
+      for (int k = 0; k < dims[2]; k++) {
+        if (vhacdVolume->GetVoxel(i, j, k) >= 2) {
+          num_filled++;
+          setBoolVoxelByIndex(Mn::Vector3i(i, j, k), true);
```

Comment:
I'm avoiding doing that right now since setBoolVoxelByIndex has to retrieve the relevant dict, hash the voxel index, and then set the boolean at that index. However I think I can do this more efficiently looking at it.